### PR TITLE
SLM-235: OTC 'code no longer valid' page and associated refactoring

### PR DESCRIPTION
**Pull request by nathanrussell-moj-digital**

### CODE REVIEW

File: server/middleware/one-time-code-auth/setUpOneTimeCode.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import RequestOneTimeCodeController from '../../routes/one-time-code-auth/Reques
 import setupOneTimeCodeEmailSent from './setupOneTimeCodeEmailSent'
 import OneTimeCodeEmailSentController from '../../routes/one-time-code-auth/OneTimeCodeEmailSentController'
 import setupVerifyOneTimeCode from './setupVerifyOneTimeCode'
-import setupStartAgain from './setupStartAgain'
```
Comment (nathanrussell-moj-digital) [inline server/middleware/one-time-code-auth/setUpOneTimeCode.ts:9]: This function simply set the route for the start again page to render the view, with no complexities such as a controller etc. I was about to create a very similar function for setting up 'code no longer valid', but decided that as both were so simple (just setting up a route to a view) I'd do them both inline

File: server/middleware/one-time-code-auth/setUpOneTimeCode.ts [hunk 2]
```diff
@@ -15,10 +15,20 @@ export default function setUpOneTimeCode(app: Express, oneTimeCodeService: OneTi
...
+    res.locals.lsjSessionDuration = `${config.lsjSessionDuration} day${config.lsjSessionDuration > 1 ? 's' : ''}`
+    next()
+  })
+
   app.use('/oneTimeCode', setUpRequestOneTimeCode(requestOneTimeCodeController))
   app.use('/oneTimeCode', setupOneTimeCodeEmailSent(oneTimeCodeEmailSentController))
-  app.use('/oneTimeCode', setupStartAgain())
   app.use('/oneTimeCode', setupVerifyOneTimeCode(verifyOneTimeCodeController))
 
+  router.get('/oneTimeCode/start-again', (req, res) => res.render('pages/one-time-code-auth/startAgain'))
```
Comment (nathanrussell-moj-digital) [inline server/middleware/one-time-code-auth/setUpOneTimeCode.ts:28]: As per comment above, setup the 2 simple routes to views

File: server/middleware/one-time-code-auth/setUpOneTimeCode.ts [hunk 3]
```diff
@@ -15,10 +15,20 @@ export default function setUpOneTimeCode(app: Express, oneTimeCodeService: OneTi
   const requestOneTimeCodeController = new RequestOneTimeCodeController(oneTimeCodeService, verifyOneTimeCodeController)
   const oneTimeCodeEmailSentController = new OneTimeCodeEmailSentController()
 
+  app.use('/oneTimeCode', (req, res, next) => {
```
Comment (nathanrussell-moj-digital) [inline server/middleware/one-time-code-auth/setUpOneTimeCode.ts:18]: A little bit of refactoring for the `oneTimeCodeValidityDuration` and `lsjSessionDuration` values.
Previously they were setup in a corresponding view class for the pages that needed them (2 places). We need the `oneTimeCodeValidityDuration` value on the 'code no longer valid' page as well, so to follow that pattern would have meant creating a third view class and a controller for 'code no longer valid' which felt like overkill.
Given that these are both readonly values and are not part of a form submission I decided to remove them from the view classes 👇 and simply add them as locals

File: server/routes/one-time-code-auth/OneTimeCodeEmailSentController.test.ts [hunk 4]
```diff
@@ -12,15 +11,9 @@ const res = {
   redirect: jest.fn(),
 }
 
-jest.mock('../../config')
-
 describe('OneTimeCodeEmailSentController', () => {
   const oneTimeCodeEmailSentController = new OneTimeCodeEmailSentController()
 
-  beforeEach(() => {
```
Comment (nathanrussell-moj-digital) [inline server/routes/one-time-code-auth/OneTimeCodeEmailSentController.test.ts:20]: `oneTimeCodeValidityDuration` is no longer part of the view 👇 so is not needed in this test

File: server/routes/one-time-code-auth/RequestOneTimeCodeController.test.ts [hunk 5]
```diff
@@ -36,11 +34,6 @@ describe('RequestOneTimeCodeController', () => {
     verifyOneTimeCodeController as unknown as VerifyOneTimeCodeController
   )
 
-  beforeEach(() => {
```
Comment (nathanrussell-moj-digital) [inline server/routes/one-time-code-auth/RequestOneTimeCodeController.test.ts:39]: `oneTimeCodeValidityDuration` and `lsjSessionDuration` are no longer part of the view 👇 so are not needed in this test

File: server/routes/one-time-code-auth/VerifyOneTimeCodeController.test.ts [hunk 6]
```diff
@@ -64,7 +64,7 @@ describe('VerifyOneTimeCodeController', () => {
       )
 
       expect(req.flash).not.toHaveBeenCalled()
-      expect(res.redirect('/barcode/find-recipient'))
```
Comment (nathanrussell-moj-digital) [inline server/routes/one-time-code-auth/VerifyOneTimeCodeController.test.ts:67]: All of these expectations on `res.redirect` were wrong (syntax), so weren't actually testing anything! Whoops! 😬

File: server/routes/one-time-code-auth/VerifyOneTimeCodeController.test.ts [hunk 7]
```diff
@@ -143,10 +143,8 @@ describe('VerifyOneTimeCodeController', () => {
         next as undefined as NextFunction
       )
 
-      expect(req.flash).toHaveBeenCalledWith('errors', [
-        { href: '#email', text: 'The code you used is no longer valid. Request a new one to sign in.' },
-      ])
-      expect(res.redirect('request-code'))
+      expect(req.flash).not.toHaveBeenCalled()
```
Comment (nathanrussell-moj-digital) [inline server/routes/one-time-code-auth/VerifyOneTimeCodeController.test.ts:146]: This is the test for the core change for this requirement

**Review by mikehalmamoj — APPROVED** [review]

**Pull request merged**